src/sentry/integrations/request_buffer.py

```diff
@@ -0,0 +1,153 @@
+    This should store the aggregate counts of each type for last 30 days for each integration
+    """
+
+    def __init__(self, key):
+        self.integrationkey = key
+        logger = logging.getLogger(__name__)
+
+        try:
+            cluster_id = settings.SENTRY_INTEGRATION_ERROR_LOG_REDIS_CLUSTER
+            self.client = redis.redis_clusters.get(cluster_id)
+        except KeyError as e:
+            logger.info("no_redis_cluster", extra={"error": e, "cluster_id": cluster_id})
+
+    def _convert_obj_to_dict(self, redis_object):
+        """
+        Convert the request string stored in Redis to a python dict
+        """
+
+        return json.loads(redis_object)
+
+    def _get_all_from_buffer(self, buffer_key):
+        """
+        Get the list at the buffer key.
+        """
+
+        return self.client.lrange(buffer_key, 0, BUFFER_SIZE - 1)
+
+    def _get_broken_range_from_buffer(self, buffer_key):
+        """
+        Get the list at the buffer key in the broken range.
+        """
+
+        return self.client.lrange(buffer_key, 0, IS_BROKEN_RANGE - 1)
+
+    def _get(self):
+        """
+        Returns the list of daily aggregate error counts.
+        """
+        return [
+            self._convert_obj_to_dict(obj)
+            for obj in self._get_broken_range_from_buffer(self.integrationkey)
+        ]
+
+    def is_integration_broken(self):
+        """
+        Integration is broken if we have 7 consecutive days of errors and no successes OR have a fatal error
+
+        """
+        items = self._get()
+
+        data = [
+            datetime.strptime(item.get("date"), "%Y-%m-%d").date()
+            for item in items
+            if item.get("fatal_count", 0) > 0 and item.get("date")
+        ]
+
+        if len(data) > 0:
+            return True
+
+        data = [
+            datetime.strptime(item.get("date"), "%Y-%m-%d").date()
+            for item in items
+            if item.get("error_count", 0) > 0
+            and item.get("success_count", 0) == 0
+            and item.get("date")
+        ]
+
+        if not len(data):
+            return False
+
+        if len(data) < IS_BROKEN_RANGE:
+            return False
+
+        date_set = {data[0] - timedelta(x) for x in range((data[0] - data[-1]).days)}
+        missing = list(date_set - set(data))
+        if len(missing):
+            return False
+        return True
+
+    def add(self, count: str):
+        VALID_KEYS = ["success", "error", "fatal"]
+        if count not in VALID_KEYS:
+            raise Exception("Requires a valid key param.")
+
+        other_count1, other_count2 = list(set(VALID_KEYS).difference([count]))[0:2]
+        now = datetime.now().strftime("%Y-%m-%d")
+
+        buffer_key = self.integrationkey
+        pipe = self.client.pipeline()
+
+        while True:
+            try:
+                pipe.watch(buffer_key)
+                recent_item_array = self.client.lrange(
```

Comment:
This could be included in the pipe, ie. `pipe.lrange(...)`. I don't believe there are any performance or correctness implications to doing it either way, just a logical/organizational thing.